src/array.c

```diff
@@ -86,7 +86,17 @@ static jl_array_t *_new_array_(jl_value_t *atype, uint32_t ndims, size_t *dims,
 
     int ndimwords = jl_array_ndimwords(ndims);
     int tsz = JL_ARRAY_ALIGN(sizeof(jl_array_t) + ndimwords*sizeof(size_t), JL_CACHE_BYTE_ALIGNMENT);
-    if (tot <= ARRAY_INLINE_NBYTES) {
+    if (elsz == 1) { // allocate byte arrays with jl_gc_alloc_buf
+        tsz = JL_ARRAY_ALIGN(tsz, JL_CACHE_BYTE_ALIGNMENT); // align whole object
+        a = (jl_array_t*)jl_gc_alloc(ptls, tsz, atype);
+        JL_GC_PUSH1(&a);
+        data = jl_gc_alloc_buf(ptls, tot);
```

Comment:
I believe the `how` needs to be initialized to `0` first before calling into the GC again so that the marking code won't try to trace a dangling pointer.